dma: stm32g4: Fix dmamux number of requests and enable additonal tests

str4t0m opened the pull request:
This PR fixes the dma-requests dt property for STM32G4 SOCs.

Additionally it adds nucleo-g474 to the uart async api, the dma/chan_blen_transfer and the dma/loop_transfer test.
The nucleo-g431 is added to the spi_loopback test, such that one g4 socs runs it in async configuration.

str4t0m committed:
dts: stm32g4: fix dmamux number of dmamux request inputs

For dmamux nodes the dma-requests property specifies the number of
peripheral request inputs(not nr. of request trigger inputs).
This commit fixes this for g4 series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: uart_async_api: add nucleo_g474re board

Adds nucleo_g474re board to the uart_async_api test.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: dma: add nucleog474re board

Add nucleo_g474re board to dma/chan_blen_transfer and dma/loop_transfer
tests.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: spi: add nucleo_g431rb to spi_loopback test in async mode

Add nucleo_g431rb to spi_loopback test using spi async configuration.
Such that tests do not only run with syncronous spi configuration
(nucleo_g474re), but also with the asyncronous configuration.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

FRASTM commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -546,7 +546,7 @@
 			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x4>;
 			dma-generators = <4>;
-			dma-requests= <21>;
+			dma-requests= <115>;
```
The Ref man of the stm32g4 (RM0440) is  giving 115 in the DMAMUX instantiation table and counting the DMAMUX request inputs from 1 to 115 , (where inputs 1 to 4 are for generator). In the dmamux bindings, the dma-request is based on the stm32wb. There (RM0434),  the DMAMUX instantiation table gives 36 peripheral request inputs : the list is counting from 1 to 40 with 4 dmamux generators. It is confusing but I guess the dma-requests is 115-4=111 (as it should be in the G4 refman). Actually, the dma-request is the nb of peripheral request inputs (excluding the generator inputs).

str4t0m replied:
Thanks, fixed that now.

FRASTM commented:
Beside this change and considering that dmamux request line Multiplexer ID (<dma_slot>) is directly given by the DMAMUX assignment table (of the stm32 ref. man.), the dmamux request_id  should not be strictly greater than  dev_config->req_nb + dev_config->gen_nb (but could be equal to) in the dmamux_stm32_configure() of the drivers/dma/dmamux_stm32.c : 
```
	if (request_id > dev_config->req_nb + dev_config->gen_nb) {
		LOG_ERR("DMA MUX request ID %d is too big.", request_id);
		return -EINVAL;
	}
```

str4t0m committed:
drivers/dma: stm32 fix dmamux request id valid check

The request id is given by the DMA request MUX id which start at offset
1 and are vaid until req_nb + gen_nb.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented:
I have updated the request_id valid check to cause an error if id is 0 or greater than the sum req_nb + gen_nb.

FRASTM commented:
excellent ! Thanks @ @str4t0m

str4t0m commented:
Rebased after merge of #34297  to resolve conflict in tests/drivers/uart/uart_async_api/src/test_uart.h.

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_g431rb.conf:
```diff
@@ -0,0 +1,12 @@
+
+# Reduce bus clock speed to be able to reach SPI_LOOPBACK_SLOW_FREQ = 500000
+# with max prescaler 256
+CONFIG_CLOCK_STM32_APB2_PRESCALER=2
```
Please convert to device tree.

str4t0m replied:
Thanks for pointing this out.
I see that the pr that changes the config to dt was not merged for nucleo_g474re.
It would probably be best to update it for nucleo_g474re as well, and block until #34709 was merged?

Not related to this pr, but there is a similar case in watchdog/wdt_basic_api, have you already taken account for that one?

erwango replied:
> Thanks for pointing this out.
> I see that the pr that changes the config to dt was not merged for nucleo_g474re.
> It would probably be best to update it for nucleo_g474re here as well, and block until #34709 was merged, or would you prefer that #34709 accounts for this itself?

Aim is to get https://github.com/zephyrproject-rtos/zephyr/pull/34709 merged today.

> Not related to this pr, but there is a similar case in watchdog/wdt_basic_api, have you already taken account for that one?

Sure, I'll take care about it.

erwango reviewed (approved):
Approve, but will require to have the previous issue fixed

str4t0m committed:
tests: spi: convert nucleo_g474re to define prescaler in dt

As the nucleo_g474re was converted to define clock properties
in devicetree, this commit makes the ncecessary changes such that
the ncecessary test properties are still applied.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented:
This pr is still block by #34758, once that pr is merged the spi_loopback on nucleo_g431rb will pass.
In the last run additionally two unrelated logging/dictionary on qemu_cortex_a53 failed.
I hope this can be finished in the 2.6.0 merge window.
